File: selectionsort/selectionsort.go
package selectionsort

func findSmallest(arr []int) int {
	smallest := arr[0]
	smallestIndex := 0

	for i, v := range arr {
		if v < smallest {
			smallest = v
			smallestIndex = i
		}
	}
	return smallestIndex
}

// O(n * n)
func selectionSort(arr []int) []int {
	newArr := []int{}

	l := len(arr)
	for i := 0; i < l; i++ {
		smallestIndex := findSmallest(arr)
		newArr = append(newArr, arr[smallestIndex])
		arr = append(arr[:smallestIndex], arr[smallestIndex+1:]...)
	}

	return newArr
}

File: SwiftAlgorithms/SwiftAlgorithms/SelectionSort/SelectionSort.swift
//
//  selectsort.swift
//  SwiftAlgorithms
//
//  Created by ke on 2024/7/2.
//

import Foundation

struct SelectionSort {
    static func findSmallest(_ arr: [Int]) -> Int {
        var smallest = arr[0]
        var smallestIndex = 0
        
        
        for (i,v) in arr.enumerated() {
            if v < smallest {
                smallest = v
                smallestIndex = i
            }
        }
        
        return smallestIndex
    }
    
    static func selectionSort(_ arr: inout [Int]) -> [Int] {
        var newArr: [Int] = []
        
        for _ in 0..<arr.count {
            let smallestIndex = findSmallest(arr)
            newArr.append(arr[smallestIndex])
            arr.remove(at: smallestIndex)
        }
        
        return newArr
    }
}
